exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/UserServer.java

```diff
@@ -328,44 +304,73 @@ protected void consumeHandshake(ChannelHandlerContext ctx, UserToBitHandshake in
           connection.disableReadTimeout();
           logger.warn("Timeout Disabled as client doesn't support it.", connection.getName());
         }
 
-        BitToUserHandshake.Builder respBuilder = BitToUserHandshake.newBuilder()
+        final BitToUserHandshake.Builder respBuilder = BitToUserHandshake.newBuilder()
             .setRpcVersion(UserRpcConfig.RPC_VERSION);
 
         try {
-          if (inbound.getRpcVersion() != UserRpcConfig.RPC_VERSION) {
-            final String errMsg = String.format("Invalid rpc version. Expected %d, actual %d.",
-                UserRpcConfig.RPC_VERSION, inbound.getRpcVersion());
+          if (!SUPPORTED_RPC_VERSIONS.contains(inbound.getRpcVersion())) {
+            final String errMsg = String.format("Invalid rpc version. Expected %s, actual %d.",
+                SUPPORTED_RPC_VERSIONS, inbound.getRpcVersion());
 
             return handleFailure(respBuilder, HandshakeStatus.RPC_VERSION_MISMATCH, errMsg, null);
           }
 
-          if (authenticator != null) {
+          connection.setHandshake(inbound);
+
+          if (authFactory == null) { // authentication is disabled
+            connection.finalizeSession(inbound.getCredentials().getUserName());
+            respBuilder.setStatus(HandshakeStatus.SUCCESS);
+            return respBuilder.build();
+          }
+
+          if (inbound.getRpcVersion() == NON_SASL_RPC_VERSION_SUPPORTED) { // for backward compatibility
+            final String userName = inbound.getCredentials().getUserName();
+            if (logger.isTraceEnabled()) {
+              logger.trace("User {} on connection {} is using an older client (Drill version <= 1.8).",
+                  userName, connection.getRemoteAddress());
+            }
             try {
               String password = "";
               final UserProperties props = inbound.getProperties();
               for (int i = 0; i < props.getPropertiesCount(); i++) {
                 Property prop = props.getProperties(i);
-                if (UserSession.PASSWORD.equalsIgnoreCase(prop.getKey())) {
+                if (ConnectionParameters.PASSWORD.equalsIgnoreCase(prop.getKey())) {
                   password = prop.getValue();
                   break;
                 }
               }
-              authenticator.authenticate(inbound.getCredentials().getUserName(), password);
+              final PlainMechanism plainMechanism = authFactory.getPlainMechanism();
+              if (plainMechanism == null) {
+                throw new UserAuthenticationException("The server no longer supports username/password" +
+                    " based authentication. Please talk to your system administrator.");
+              }
+              plainMechanism.getAuthenticator()
+                  .authenticate(userName, password);
+              connection.changeHandlerTo(handler);
+              connection.finalizeSession(userName);
+              respBuilder.setStatus(HandshakeStatus.SUCCESS);
```

Comment:
should the rpc version of the response be changed to `NON_SASL_RPC_VERSION_SUPPORTED` in that specific case?